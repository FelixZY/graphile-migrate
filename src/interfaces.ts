/** Represents the graphile_migrate.current type in the DB */
export interface DbCurrent {
  filename: string;
  content: string;
  date: Date;
}
