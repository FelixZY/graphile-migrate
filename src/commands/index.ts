export { commit } from "./commit";
export { compile } from "./compile";
export { init } from "./init";
export { migrate } from "./migrate";
export { reset } from "./reset";
export { run } from "./run";
export { status } from "./status";
export { uncommit } from "./uncommit";
export { watch } from "./watch";
