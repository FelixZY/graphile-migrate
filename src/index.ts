export * from "./commands";
export { defaultLogger } from "./logger";
export { Settings } from "./settings";
export { LogFunctionFactory, Logger, LogLevel } from "@graphile/logger";
