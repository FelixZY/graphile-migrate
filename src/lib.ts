export function mergeWithoutClobbering(
  original: { [key: string]: string | undefined },
  newStuff: { [key: string]: string | undefined },
  message: string,
): { [key: string]: string | undefined } {
  const result = { ...original };
  for (const key in newStuff) {
    if (typeof result[key] === "undefined" || result[key] === newStuff[key]) {
      result[key] = newStuff[key];
    } else {
      throw new Error(
        `Refusing to clobber '${key}' (from '${original[key]}' to '${newStuff[key]}'): ${message}`,
      );
    }
  }

  return result;
}

export function isLoggedError(error: unknown): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "_gmlogged" in error &&
    error._gmlogged === true
  );
}

export function errorCode(e: unknown): string | null {
  return typeof e === "object" &&
    e !== null &&
    "code" in e &&
    typeof e.code === "string"
    ? e.code
    : null;
}
