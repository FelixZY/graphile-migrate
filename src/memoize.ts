/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-assignment */
export default function memoize<T extends (...args: Array<any>) => any>(
  fn: T,
): (...funcArgs: Parameters<T>) => ReturnType<T> {
  let lastArgs: Parameters<T>;
  let lastResult: ReturnType<T>;
  return (...args: Parameters<T>): ReturnType<T> => {
    if (
      lastArgs &&
      args.length === lastArgs.length &&
      args.every((arg, i) => arg === lastArgs[i])
    ) {
      return lastResult;
    } else {
      lastArgs = args;
      lastResult = fn(...args);
      return lastResult;
    }
  };
}
