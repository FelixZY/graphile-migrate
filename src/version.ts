// This file is autogenerated by /scripts/version.mjs
export const version = "2.0.0-rc.1";
