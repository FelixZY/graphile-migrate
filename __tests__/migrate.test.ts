import "./helpers"; // Has side-effects; must come first

import mockFs from "mock-fs";

import { migrate } from "../src";
import { withClient } from "../src/pg";
import { ParsedSettings, parseSettings } from "../src/settings";
import { makeMigrations, resetDb, settings } from "./helpers";

beforeEach(resetDb);
beforeEach(async () => {
  mockFs({ migrations: mockFs.directory() });
});
afterEach(() => {
  mockFs.restore();
});
const {
  MIGRATION_1_TEXT,
  MIGRATION_1_COMMITTED,
  MIGRATION_ENUM_COMMITTED,
  MIGRATION_NOTRX_TEXT,
  MIGRATION_NOTRX_COMMITTED,
} = makeMigrations();

function getStuff(parsedSettings: ParsedSettings) {
  return withClient(
    parsedSettings.connectionString,
    parsedSettings,
    async (pgClient, _context) => {
      const { rows: migrations } = await pgClient.query(
        "select * from graphile_migrate.migrations",
      );
      const { rows: tables } = await pgClient.query(
        "select * from pg_class where relnamespace = 'public'::regnamespace and relkind = 'r'",
      );
      const { rows: enums } = await pgClient.query(
        "select typname, (select count(*) from pg_enum where enumtypid = pg_type.oid) as value_count from pg_type where typnamespace = 'public'::regnamespace and typtype = 'e'",
      );
      return { migrations, tables, enums };
    },
  );
}

it("runs migrations", async () => {
  mockFs({
    "migrations/current.sql": MIGRATION_1_TEXT,
  });

  await migrate(settings);

  const parsedSettings = await parseSettings(settings);

  {
    const { migrations, tables, enums } = await getStuff(parsedSettings);
    expect(migrations).toHaveLength(0);
    expect(tables).toHaveLength(0);
    expect(enums).toHaveLength(0);
  }

  mockFs({
    [`migrations/committed/000001.sql`]: MIGRATION_1_COMMITTED,
    [`migrations/committed/000002.sql`]: MIGRATION_ENUM_COMMITTED,
    "migrations/current.sql": MIGRATION_NOTRX_TEXT,
  });

  await migrate(settings);

  {
    const { migrations, tables, enums } = await getStuff(parsedSettings);

    expect(migrations).toHaveLength(2);
    expect(migrations.map(({ date, ...rest }) => rest)).toMatchSnapshot();
    expect(tables).toHaveLength(1);
    expect(tables.map((t) => t.relname)).toMatchSnapshot();
    expect(enums).toHaveLength(1);
    expect(enums).toMatchSnapshot();
  }

  mockFs({
    [`migrations/committed/000001.sql`]: MIGRATION_1_COMMITTED,
    [`migrations/committed/000002.sql`]: MIGRATION_ENUM_COMMITTED,
    [`migrations/committed/000003.sql`]: MIGRATION_NOTRX_COMMITTED,
    "migrations/current.sql": "",
  });

  await migrate(settings);

  {
    const { migrations, tables, enums } = await getStuff(parsedSettings);

    expect(migrations).toHaveLength(3);
    const mappedMigrations = migrations.map(({ date, ...rest }) => rest);
    expect(mappedMigrations).toMatchSnapshot();
    expect(tables).toHaveLength(1);
    const mappedTables = tables.map((t) => t.relname);
    expect(mappedTables).toMatchSnapshot();
    expect(enums).toHaveLength(1);
    expect(enums).toMatchSnapshot();
  }
});

it("refuses to run migration with invalid hash", async () => {
  mockFs({
    [`migrations/committed/000001.sql`]: MIGRATION_1_COMMITTED,
    [`migrations/committed/000002.sql`]:
      MIGRATION_ENUM_COMMITTED +
      "\ncomment on type user_role is 'this invalidates the hash';",
    [`migrations/committed/000003.sql`]: MIGRATION_NOTRX_COMMITTED,
    "migrations/current.sql": "",
  });

  await expect(migrate(settings)).rejects.toThrowErrorMatchingSnapshot();
});

it("will run a migration with invalid hash if told to do so", async () => {
  const parsedSettings = await parseSettings(settings);

  mockFs({
    [`migrations/committed/000001.sql`]: MIGRATION_1_COMMITTED,
    [`migrations/committed/000002.sql`]:
      "--! AllowInvalidHash\n" +
      MIGRATION_ENUM_COMMITTED +
      "\ncomment on type user_role is 'this invalidates the hash';",
    [`migrations/committed/000003.sql`]: MIGRATION_NOTRX_COMMITTED,
    "migrations/current.sql": "",
  });

  await migrate(settings);

  {
    const { migrations, enums } = await getStuff(parsedSettings);

    expect(migrations).toHaveLength(3);
    const mappedMigrations = migrations.map(({ date, ...rest }) => rest);
    expect(mappedMigrations).toMatchSnapshot();
    expect(enums).toHaveLength(1);
    expect(enums).toMatchSnapshot();
  }
});
