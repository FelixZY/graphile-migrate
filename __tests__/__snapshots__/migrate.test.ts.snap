// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles errors during migration gracefully 1`] = `[error: division by zero]`;

exports[`handles errors during migration gracefully 2`] = `
"
[31m[1mðŸ›‘ Error occurred whilst processing migration[22m[39m
    error: division by zero
        at ~/node_modules/pg/lib/client.js:[LINE]:[COL]
        at processTicksAndRejections (node:internal/process/task_queues:[LINE]:[COL])
        at runQueryWithErrorInstrumentation (~/src/instrumentation.ts:[LINE]:[COL])
        at ~/src/migration.ts:[LINE]:[COL]
        at withAdvisoryLock (~/src/pgReal.ts:[LINE]:[COL])
        at runCommittedMigration (~/src/migration.ts:[LINE]:[COL])
        at ~/src/commands/migrate.ts:[LINE]:[COL]
        at withAdvisoryLock (~/src/pgReal.ts:[LINE]:[COL])
        at ~/src/commands/migrate.ts:[LINE]:[COL]
        at withClient (~/src/pgReal.ts:[LINE]:[COL])
        at _migrate (~/src/commands/migrate.ts:[LINE]:[COL])
        at Object.<anonymous> (~/__tests__/migrate.test.ts:[LINE]:[COL])

    Severity:	ERROR
    Code:    	22012
,[object Object]"
`;

exports[`refuses to run migration with invalid hash 1`] = `"Hash for 000002.sql does not match - sha1:cbed240dda7dfa510ff785783bbe6af7743b3a11 !== sha1:bddc1ead3310dc1c42cdc7f63537ebdff2e9fd7b; has the file been tampered with?"`;

exports[`runs migrations 1`] = `
[
  {
    "filename": "000001.sql",
    "hash": "sha1:e00ec93314a423ee5cc68d1182ad52f16442d7df",
    "previous_hash": null,
  },
  {
    "filename": "000002.sql",
    "hash": "sha1:bddc1ead3310dc1c42cdc7f63537ebdff2e9fd7b",
    "previous_hash": "sha1:e00ec93314a423ee5cc68d1182ad52f16442d7df",
  },
]
`;

exports[`runs migrations 2`] = `
[
  "foo",
]
`;

exports[`runs migrations 3`] = `
[
  {
    "typname": "user_role",
    "value_count": "1",
  },
]
`;

exports[`runs migrations 4`] = `
[
  {
    "filename": "000001.sql",
    "hash": "sha1:e00ec93314a423ee5cc68d1182ad52f16442d7df",
    "previous_hash": null,
  },
  {
    "filename": "000002.sql",
    "hash": "sha1:bddc1ead3310dc1c42cdc7f63537ebdff2e9fd7b",
    "previous_hash": "sha1:e00ec93314a423ee5cc68d1182ad52f16442d7df",
  },
  {
    "filename": "000003.sql",
    "hash": "sha1:2d248344ac299ebbad2aeba5bfec2ae3c3cb0a4f",
    "previous_hash": "sha1:bddc1ead3310dc1c42cdc7f63537ebdff2e9fd7b",
  },
]
`;

exports[`runs migrations 5`] = `
[
  "foo",
]
`;

exports[`runs migrations 6`] = `
[
  {
    "typname": "user_role",
    "value_count": "2",
  },
]
`;

exports[`will run a migration with invalid hash if told to do so 1`] = `
[
  {
    "filename": "000001.sql",
    "hash": "sha1:e00ec93314a423ee5cc68d1182ad52f16442d7df",
    "previous_hash": null,
  },
  {
    "filename": "000002.sql",
    "hash": "sha1:bddc1ead3310dc1c42cdc7f63537ebdff2e9fd7b",
    "previous_hash": "sha1:e00ec93314a423ee5cc68d1182ad52f16442d7df",
  },
  {
    "filename": "000003.sql",
    "hash": "sha1:2d248344ac299ebbad2aeba5bfec2ae3c3cb0a4f",
    "previous_hash": "sha1:bddc1ead3310dc1c42cdc7f63537ebdff2e9fd7b",
  },
]
`;

exports[`will run a migration with invalid hash if told to do so 2`] = `
[
  {
    "typname": "user_role",
    "value_count": "2",
  },
]
`;
